### PR TITLE
[tools] Added a tool to generate the Golang dependency tree

### PR DESCRIPTION
**Pull request by sgnn7**
### What does this PR do?

Since Golang CLI and built-in tooling was inadequate for this task, we
now have our own script/tool that can generate a dependency tree as
close to the one that Golang will actually use as we can.

Usage:
```sh-session
$ inv agent.build-dep-tree
or
$ inv agent.build-dep-tree --git-ref=<tag>    # Will only work for tags created after the merge of this PR
```

### Motivation

Need for an accurate Golang dependency tree (AC-791)

### Additional Notes

Anything else we should know when reviewing?

### Describe how to test your changes

N/A

### CODE REVIEW

File: tasks/agent.py [hunk 1]
```diff
@@ -534,6 +534,24 @@ def omnibus_build(
     print("Omnibus: {}".format(omnibus_elapsed))
 
 
+@task
+def build_dep_tree(
+    ctx, tag="",
```
Comment (mx-psi) [inline tasks/agent.py:539]: Maybe reword as `git_ref`?
```suggestion
    ctx, git_ref="",
```

You can actually pass any git reference to `git checkout` and I can see someone wanting to switch to a branch/commit instead of a tag.

**Review by mx-psi — CHANGES_REQUESTED** [review]
Thanks for doing this! This is something I have needed recently and I did not find any tool that did this properly so it's good to have one 😄 

I left some comments, mostly minor code fixes so that we can maintain this more easily. My biggest concern is that it's not super easy to `git checkout` to a given ref if we host this on the Datadog Agent repository and I think that is something we will probably want to do. I would suggest hosting it in a separate repository and handle it as just another Go tool in our tooling, but other alternatives are also possible.

I might do a second pass later but I wanted to give you some feedback quickly

File: tools/dep_tree_resolver/go_deps.go [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+	if Debug {
+		fmt.Printf("- Found %d dep(s) for module %s\n", len(modDeps), actualModule)
+	}
+
+	dependencies := []DependencyTree{}
+	for _, depModule := range modDeps {
+		// Circular dependencies are ignored once we encounter a loop
+		if isCirccularDependency(&depModule, &rootDepPath) {
+			if Debug {
+				fmt.Printf("WARN: Circular dependency breakout for %s\n", depModule.Path)
```
[the remark this answers is not in the document]
Comment (sgnn7) [reply tools/dep_tree_resolver/go_deps.go:301]: The warning is just for debug level. It's not normally visible and is pretty valuable in figuring out if there's something wrong with the logic since the whole tree under the exclusion gets omitted.

File: tools/dep_tree_resolver/go_deps.go [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+// to a file or stdout
+const WriteToFile = true
+
+// OutputFileName is our target output filename for the generated
+// dependency tree
+const OutputFileName = "dependency_tree.txt"
+
+// RootModulePlaceholderVersion denotes what version we will assign to our
+// root module as generally we won't have one provided by the Golang tooling
+const RootModulePlaceholderVersion = "master"
```
[the remark this answers is not in the document]
Comment (sgnn7) [reply tools/dep_tree_resolver/go_deps.go:28]: That's doable but then we have to depend on `git` to tell us that info. As-is the code works on any (git and non-git) repo. Maybe just the name can be changed to something more representative of the version we are looking at (e.g. `current`)?

[the remark this answers is not in the document]
Comment (mx-psi) [reply tools/dep_tree_resolver/go_deps.go:28]: `current` works for me!

**Comment by sgnn7** [comment]
Hey @mx-psi, 

Thanks for the feedback - it's much appreciated and I'll see about fixing your comments this week!

> My biggest concern is that it's not super easy to git checkout to a given ref if we host this on the Datadog Agent repository and I think that is something we will probably want to do. I would suggest hosting it in a separate repository and handle it as just another Go tool in our tooling, but other alternatives are also possible.

I talked to @olivielpeau about this and we both are leaning towards leaving this in-repo until maybe we have a clear understanding of what we may want to do with this tool as it currently may be a very low payoff for a lot of work. At this time the only user of this output is the security team and if they require running this against an older version of the codebase, it would be a one-time thing and on the agent-core team anyways. Let me know what your thoughts are on this and then we can figure out what our next steps would be. 

Thank you for your time taking a look at this!

**Comment by mx-psi** [comment]
> I talked to @olivielpeau about this and we both are leaning towards leaving this in-repo until maybe we have a clear understanding of what we may want to do with this tool as it currently may be a very low payoff for a lot of work. At this time the only user of this output is the security team and if they require running this against an older version of the codebase, it would be a one-time thing and on the agent-core team anyways.

Besides the issue about not being able to run the script against older codebases, the other issue I see with the current approach is that when running the script against a given tag/git ref we would not have any bugfixes we added after that reference. I think that would be very confusing and I would maybe suggest removing the `git_ref` argument altogether if we are not going to handle that, so that the user is forced to think about it.

>  it currently may be a very low payoff for a lot of work

What I suggest would entail copying the `tools/dep_tree_resolver` folder to a new repository, say `github.com/DataDog/datadog-agent-tools` and replacing `go run tools/dep_tree_resolver/go_deps.go` by `go run github.com/DataDog/datadog-agent-tools/dep_tree_resolver`. That should work exactly as it does now except that it would fetch the latest version from Github. I don't think that would be a lot of additional work but it's also something we can deal with later if it is not a concern.

File: tools/dep_tree_resolver/go_deps.go [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+// childern in a recursive tree
+type DependencyTree struct {
+	Mod          Module
+	Dependencies []DependencyTree
+}
+
+func runCommand(ctx context.Context, executable string, command ...string) ([]byte, error) {
+	cmd := exec.CommandContext(ctx, executable, command...)
+	outputBuf := bytes.NewBuffer(nil)
+	cmd.Stderr = outputBuf
```
[the remark this answers is not in the document]
Comment (sgnn7) [reply tools/dep_tree_resolver/go_deps.go:54]: If we don't do that, the error output text won't be shown. This way both stdout and stderr end up in the same buffer. We could do more complex logic around this to isolate and print each individually but it really didn't seem needed.

File: tools/dep_tree_resolver/go_deps.go [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+	}
+
+	io.WriteString(buf, fmt.Sprintf("- %s@%s\n", (*depTree).Mod.Path, (*depTree).Mod.Version))
+	for _, depSubtree := range (*depTree).Dependencies {
+		printDepTree(buf, &depSubtree, level+1, skipDuplicates)
+	}
+}
+
+func main() {
+	ctx, _ := context.WithCancel(context.Background())
```
[the remark this answers is not in the document]
Comment (sgnn7) [reply tools/dep_tree_resolver/go_deps.go:373]: Yeah I over-engineered this early to support cancelling during parallel module resolution but since that's not used, I'll go with your suggestion.

File: tools/dep_tree_resolver/go_deps.go [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+}
+
+// TODO: Actually use `skipDuplicates` value
+func printDepTree(buf *bufio.Writer, depTree *DependencyTree, level int, skipDuplicates bool) {
+	for idx := 0; idx < level; idx++ {
+		io.WriteString(buf, "    ")
+	}
+
+	io.WriteString(buf, fmt.Sprintf("- %s@%s\n", (*depTree).Mod.Path, (*depTree).Mod.Version))
+	for _, depSubtree := range (*depTree).Dependencies {
```
[the remark this answers is not in the document]
Comment (sgnn7) [reply tools/dep_tree_resolver/go_deps.go:367]: The deref for this particular line makes sense but passing around structs I think would:
- Slow down the already slow code
- Make it jump exponentially in memory usage as it would use copies of same structs (and we have a very large tree to work with)
- Would possibly cause unintended issues with modifying a module object within a method only to find out that the parent caller didn't have its sent arg changed as expected

**Comment by sgnn7** [comment]
Hey @mx-psi,
I fixed up most of your comments so feel free to re-review when you get a chance. Since the core team is still unsure if this will see any real-life use yet (even security team might not use this in the future) we will add this directly into the repo for now. If we start using the tool more widely, we can easily split it out to a different repo and treat it as a real tool dependency it's just that the needed effort at this time really doesn't seem warranted.

File: tasks/agent.py [hunk 7]
```diff
@@ -534,6 +534,32 @@ def omnibus_build(
...
+        saved_branch = result.stdout
+
+        ctx.run("git checkout {}".format(git_ref))
+    else:
+        print("No tag specified. Using the current state of repository.")
+
+    ctx.run("go run tools/dep_tree_resolver/go_deps.go")
+
+    if saved_branch:
+        ctx.run("git checkout {}".format(saved_branch), hide='stdout')
```
Comment (mx-psi) [inline tasks/agent.py:560]: nit
```suggestion
    try:
        ctx.run("go run tools/dep_tree_resolver/go_deps.go")
    finally:
        if saved_branch:
            ctx.run("git checkout {}".format(saved_branch), hide='stdout')
```

so that we restore the saved branch regardless of failure

**Review by mx-psi — APPROVED** [review]
Looks good, just a minor nit. And by the way, you are going to have an user for sure, me, since I need to hunt down some dependency issues for OpenTelemetry :)

**Review by truthbk — APPROVED** [review]
